cartographer_ros/cartographer_ros/map_builder_bridge.cc

```diff
@@ -246,6 +247,97 @@ visualization_msgs::MarkerArray MapBuilderBridge::GetTrajectoryNodesList() {
   return trajectory_nodes_list;
 }
 
+visualization_msgs::MarkerArray MapBuilderBridge::GetConstraintsList(
+    const ::cartographer::mapping::SparsePoseGraph::Constraint::Tag
+        wanted_constraint_tag) {
+  visualization_msgs::MarkerArray constraints_viz;
+
+  const auto trajectory_nodes =
+      map_builder_.sparse_pose_graph()->GetTrajectoryNodes();
+  const auto constraints = map_builder_.sparse_pose_graph()->constraints();
+  std::vector<cartographer::transform::Rigid3d> submap_transforms;
+  for (auto& single_trajectory : trajectory_nodes) {
+    for (auto& node : single_trajectory) {
+      auto current_trajectory_transforms =
+          map_builder_.sparse_pose_graph()->GetSubmapTransforms(
+              node.constant_data->trajectory_id);
+      std::move(current_trajectory_transforms.begin(),
+                current_trajectory_transforms.end(),
+                std::back_inserter(submap_transforms));
+    }
+  }
+
+  int marker_id = 0;
+  ros::Time now = ros::Time::now();
+  for (const auto& constraint : constraints) {
+    visualization_msgs::Marker constraint_marker, residual_error_marker;
+
+    std_msgs::ColorRGBA color_constraint, color_error;
+
+    if (constraint.tag != wanted_constraint_tag) {
+      continue;
+    }
+
+    if (constraint.tag ==
+        cartographer::mapping::SparsePoseGraph::Constraint::INTER_SUBMAP) {
+      // yellow
+      color_constraint.a = 1.0;
+      color_constraint.r = color_constraint.g = 1.0;
+      // cyan
+      color_error.a = 1.0;
+      color_error.b = color_error.g = 1.0;
+    } else {
+      color_constraint.a = color_error.a = 1.0;
+      color_constraint.g = 1.0;
+      color_error.b = 1.0;
+    }
+
+    constraint_marker = createVisualizationMarker(
```

Comment:
I agree on the first part.
But I wouldn't merge the constraint markers and the residual markers into one marker, since then you couldn't disable them individually. Or am I mistaken about this?